samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -0,0 +1,265 @@
+   * @param ssp The {@link SystemStreamPartition} to fetch the {@link Startpoint} for.
+   * @param taskName The {@link TaskName} to fetch the {@link Startpoint} for.
+   * @return {@link Startpoint} for the {@link SystemStreamPartition}, or null if it does not exist or if it is too stale.
+   */
+  public Startpoint readStartpointForTask(SystemStreamPartition ssp, TaskName taskName) {
+    Preconditions.checkState(started, "Underlying metadata store not available");
+    Preconditions.checkNotNull(ssp, "SystemStreamPartition cannot be null");
+
+    byte[] startpointBytes = metadataStore.get(toStoreKey(ssp, taskName));
+
+    if (Objects.nonNull(startpointBytes)) {
+      Startpoint startpoint = startpointSerde.fromBytes(startpointBytes);
+      if (Instant.now().minus(expirationDuration).isBefore(Instant.ofEpochMilli(startpoint.getCreatedTimestamp()))) {
+        return startpoint; // return if deserializable and if not stale
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Deletes the {@link Startpoint} for a {@link SystemStreamPartition}
+   * @param ssp The {@link SystemStreamPartition} to delete the {@link Startpoint} for.
+   */
+  public void deleteStartpoint(SystemStreamPartition ssp) {
+    deleteStartpointForTask(ssp, null);
+  }
+
+  /**
+   * Deletes the {@link Startpoint} for a {@link SystemStreamPartition} and {@link TaskName}.
+   * @param ssp ssp The {@link SystemStreamPartition} to delete the {@link Startpoint} for.
+   * @param taskName ssp The {@link TaskName} to delete the {@link Startpoint} for.
+   */
+  public void deleteStartpointForTask(SystemStreamPartition ssp, TaskName taskName) {
+    Preconditions.checkState(started, "Underlying metadata store not available");
+    Preconditions.checkNotNull(ssp, "SystemStreamPartition cannot be null");
+
+    metadataStore.delete(toStoreKey(ssp, taskName));
+  }
+
+  /**
+   * For {@link Startpoint}s keyed only by {@link SystemStreamPartition}, this method remaps the Startpoints for the specified
+   * SystemStreamPartition to SystemStreamPartition+{@link TaskName} for all tasks provided by the {@link JobModel}
+   * This method is not atomic or thread-safe. The intent is for the Samza Processor's coordinator to use this
+   * method to assign the Startpoints to the appropriate tasks.
+   * @param ssp The {@link SystemStreamPartition} that a {@link Startpoint} is keyed to.
+   * @param jobModel The {@link JobModel} is used to determine which {@link TaskName} each {@link SystemStreamPartition} maps to.
+   * @return The list of {@link TaskName}s mapped to the ssp.
+   */
+  public Set<TaskName> groupStartpointsPerTask(SystemStreamPartition ssp, JobModel jobModel) {
+    Preconditions.checkState(started, "Underlying metadata store not available");
+    Preconditions.checkNotNull(ssp, "SystemStreamPartition cannot be null");
+    Preconditions.checkNotNull(jobModel, "JobModel cannot be null");
+
+    Startpoint startpoint = readStartpoint(ssp);
```

Comment:
Instead of reading from an existing startpoint and rewriting the entry, would it be possible to make an API which just takes in the startpoint to expand and then directly write the expanded entries? Then we could do everything in a single step instead of what looks like two steps.

Reply:
That's helpful clarification. Do you have a use case for why the expansion needs to be done after the job starts up instead of when the Startpoint is written?